### PR TITLE
[scripts][builder] Move repo.path parse logic from prepare_kb to sc-builder, remove prepare_kb and optimize sc-builder

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [ ] Update documentation

This pull request optimizes the preparation of the knowledge base before building, and also transfers the prepare_kb.py logic to sc-builder. Now the knowledge base is not copied to translate .gwf and .scsi sources into .scs, this logic is performed when the files are bypassed by the sc-builder. In the future, it will be necessary to get rid of the python script call and rewrite the gwf-translator in C++. Also, all the logic of working with repo.path has been moved to sc-builder and covered with tests.

Now you no need to wait a long time when kb build in Windows.

### CODE REVIEW

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
oops

File: sc-tools/sc-builder/src/builder.cpp [hunk 1]
```diff
@@ -5,281 +5,133 @@
...
 
 Builder::Builder() = default;
 
 bool Builder::Run(BuilderParams const & params, sc_memory_params const & memoryParams)
 {
   m_params = params;
 
-  CollectFiles();
-
   // initialize sc-memory
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/src/builder.cpp:21]: This comments looks useless

File: sc-tools/sc-builder/src/builder.cpp [hunk 2]
```diff
@@ -5,281 +5,133 @@
...
-  bool noErrors = true;
-  impl::ExtParser extParser;
-  extParser(m_params.m_enabledExtPath);
-
   ScMemory::Initialize(memoryParams);
+
+  ScRepoPathCollector::Sources excludedSources, checkSources;
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/src/builder.cpp:24]: Accotding to the codestyle you should split it to the 2 lines

File: sc-tools/sc-builder/src/builder_defines.hpp [hunk 3]
```diff
@@ -0,0 +1 @@
+#define SC_PREPARE_BUILD_SCRIPTS_PATH "/home/nikita/apps/ostis/ostis-ai/ostis-web-platform/sc-machine/scripts"
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/src/builder_defines.hpp:1]: Do not commit absolute path

File: sc-tools/sc-builder/src/sc_repo_path_collector.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+#include <regex>
+#include <filesystem>
+
+#include "sc-memory/sc_utils.hpp"
+
+std::unordered_set<std::string> const ScRepoPathCollector::m_supportedSourcesFormats = {"scs", "gwf"};
+std::unordered_set<std::string> const ScRepoPathCollector::m_supportedRepoPathFormats = {"path"};
+
+namespace impl
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/src/sc_repo_path_collector.cpp:17]: why do you need this namespace

File: sc-tools/sc-builder/src/scs_translator.cpp [hunk 5]
```diff
@@ -79,9 +73,7 @@ bool SCsTranslator::TranslateImpl(Params const & params)
   SCsHelper scs(m_ctx, std::make_shared<impl::FileProvider>(params.m_fileName));
 
   if (!scs.GenerateBySCsText(data, params.m_outputStructure))
-  {
     SC_THROW_EXCEPTION(utils::ExceptionParseError, scs.GetLastError());
-  }
 
   return true;
 }
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/src/scs_translator.cpp:79]: endline

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Comment by MksmOrlov** [comment]
Can we create ./build_kb script in sc-machine and call it from other projects? Now all ./build_kb scripts in other projects are failed

File: sc-tools/sc-builder/src/scs_translator.cpp (hunk 5 above)
Comment (smousy) [inline sc-tools/sc-builder/src/scs_translator.cpp:79]: please add newline

File: sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample1.scs
Comment (smousy) [inline sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample1.scs:1]: why is it empty?

File: sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample2.scs
Comment (smousy) [inline sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample2.scs:1]: why is it empty?

File: sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample3.scs
Comment (smousy) [inline sc-tools/sc-builder/tests/repos/kb/child_repos/child_kb/samples/sample3.scs:1]: why is it empty?

File: README.md [hunk 6]
```diff
@@ -81,7 +81,7 @@ SC-machine is a core of any OSTIS-system, so you can use a reference system name
 
   ```sh
   #build kb
-  python3 scripts/build_kb.py -c ./sc-machine.ini -b bin <path to kb folder with SCs and SCg sources (or path to repo.path file)>
+  ./bin/sc-builder -f --clear -c ./sc-machine.ini -i <path to kb folder with SCs and SCg sources (or path to repo.path file)> -o <output path>
```
Comment (smousy) [inline README.md:84]: please create pr to ostis-web-platform and change build_kb.sh script

**Review by smousy — CHANGES_REQUESTED** [review]

File: sc-tools/sc-builder/tests/units/test_config.cpp
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp:1]: It is an old problem. All tests without license

File: sc-tools/sc-builder/tests/units/test_config.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  EXPECT_TRUE(builder.Run(params, memoryConfig.GetParams()));
+
+  ScParams scParams{options, {}};
+  if (configFile.IsValid())
+  {
+    ScConfigGroup group = configFile[memoryGroupName];
+    for (auto const & key : *group)
+    {
+      std::string const & value = group[key];
+      std::stringstream stream;
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp:40]: Why you would to change this?

File: sc-tools/sc-builder/tests/units/test_config.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  ScMemoryConfig memoryConfig{configFile, memoryParams};
+
+  Builder builder;
+  EXPECT_TRUE(builder.Run(params, memoryConfig.GetParams()));
+
+  ScParams scParams{options, {}};
+  if (configFile.IsValid())
+  {
+    ScConfigGroup group = configFile[memoryGroupName];
+    for (auto const & key : *group)
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp:37]: Why you would to change this?

File: sc-tools/sc-builder/tests/units/test_config.cpp
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp:1]: I fixed it for sc-builder tests

File: sc-tools/sc-builder/tests/units/test_config.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp:46]: This code checks atributes getting from config

File: sc-tools/sc-builder/tests/repos/kb/templates/template3.scs
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/repos/kb/templates/template3.scs:1]: I check repo.path parsing. I need this files to check that repo.path parser works good

File: sc-tools/sc-builder/src/sc_repo_path_collector.cpp (hunk 4 above)
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/sc_repo_path_collector.cpp:17]: To private implementation. It must be used only in this file

**Comment by NikitaZotov** [comment]
> Can we create ./build_kb script in sc-machine and call it from other projects? Now all ./build_kb scripts in other projects are failed

We need to change args for sc-builder, because bin directory can be placed differently. I changed script in platform repo https://github.com/ostis-ai/ostis-web-platform/pull/78

File: sc-tools/sc-builder/tests/units/test_config.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MksmOrlov) [reply sc-tools/sc-builder/tests/units/test_config.cpp]: you do not need auto for primitives and strings

File: sc-tools/sc-builder/tests/units/test_config.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (MksmOrlov) [reply sc-tools/sc-builder/tests/units/test_config.cpp]: stringstream does nothing

File: sc-tools/sc-builder/src/builder.cpp (hunk 2 above)
Comment (MksmOrlov) [reply sc-tools/sc-builder/src/builder.cpp]: ?

File: sc-tools/sc-builder/src/sc_repo_path_collector.cpp (hunk 4 above)
Comment (MksmOrlov) [reply sc-tools/sc-builder/src/sc_repo_path_collector.cpp:17]: make private method

File: sc-tools/sc-builder/src/builder.cpp (hunk 2 above)
Comment (NikitaZotov) [reply sc-tools/sc-builder/src/builder.cpp]: It's fixed

File: sc-tools/sc-builder/tests/units/test_config.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp]: Fixed

File: sc-tools/sc-builder/tests/units/test_config.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-builder/tests/units/test_config.cpp]: Fixed

File: sc-tools/sc-builder/tests/units/test_repo_path_collector_api.cpp [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+{
+  EXPECT_TRUE(collector.IsRepoPathFile(TEST_SIMPLE_REPO));
+  EXPECT_FALSE(collector.IsRepoPathFile(TEST_INVALID_REPO_PATH));
+}
+
+TEST_F(ScRepoPathCollectorTestAPI, SimpleRepo)
+{
+  std::string const & repoPath = TEST_SIMPLE_REPO;
+
+  ScRepoPathCollector::Sources excludedSources, checkSources;
```
Comment (MksmOrlov) [inline sc-tools/sc-builder/tests/units/test_repo_path_collector_api.cpp:54]: Accotding to the codestyle you should split it to the 2 lines

**Review by MksmOrlov — APPROVED** [review]

File: sc-tools/sc-builder/src/scs_translator.cpp (hunk 5 above)
Comment (smousy) [reply sc-tools/sc-builder/src/scs_translator.cpp]: image

**Review by smousy — APPROVED** [review]

**Pull request merged**